futures-executor/src/task_runner.rs

```diff
@@ -0,0 +1,301 @@
+use std::prelude::v1::*;
+
+use std::cell::{Cell, RefCell};
+use std::fmt;
+use std::ptr;
+use std::rc::{Rc, Weak};
+
+use futures_core::{Future, Poll, Async};
+use futures_core::task::{self, Spawn, NotifyHandle};
+use futures_util::FutureExt;
+use futures_util::stream::FuturesUnordered;
+
+/// An object for cooperatively scheduling multiple futures on one thread.
+///
+/// A `TaskRunner` provides two implementors of the `Executor` trait, allowing
+/// futures to be spawned into a `TaskRunner` and scheduled cooperatively one
+/// another on the same thread. The `CurrentThread` type, for example,
+/// internally uses a `TaskRunner` for running futures.
+///
+/// A `TaskRunner` is *not* a `Send` structure. Once created it cannot move
+/// across threads, nor can its executor handle (`TaskExecutor`).
+///
+/// Note that `TaskRunner` is likely a low-level detail you shouldn't concern
+/// yourself with, when in doubt use `CurrentThread` instead. If you're looking
+/// to spawn a list of futures within a future you're likely going to want
+/// `FuturesUnordered`. Using `TaskRunner` typically implies that you're
+/// integraing the `futures` crate into a foreign event loop (aka not Tokio
+/// which already knows about `futures`).
+pub struct TaskRunner {
+    inner: Rc<Inner>,
+    futures: Spawn<FuturesUnordered<SpawnedFuture>>,
```

Comment:
Related: I opened https://github.com/rust-lang-nursery/futures-rs/issues/735 to track this, but is there a reason why you have to box futures to put them in `FuturesUnordered` (which puts them in an `Arc` internally)? Should `FuturesUnordered` just allow `F: ?Sized`?

Reply:
Nah for now that's just how it is, but I think it'd be quite plausible for us to change it!